storage/watcher_test.go

```diff
@@ -73,6 +74,75 @@ func TestWatcherWatchID(t *testing.T) {
 	}
 }
 
+// TestWatcherWatchPrefix tests if Watch operation correctly watches
+// and returns events with matching prefixes.
+func TestWatcherWatchPrefix(t *testing.T) {
+	b, tmpPath := backend.NewDefaultTmpBackend()
+	s := WatchableKV(newWatchableStore(b, &lease.FakeLessor{}))
+	defer cleanup(s, b, tmpPath)
+
+	w := s.NewWatchStream()
+	defer w.Close()
+
+	idm := make(map[WatchID]struct{})
+
+	prefixMatch := true
+	val := []byte("bar")
+	key, keyL := []byte("foo"), []byte("foobar")
+
+	for i := 0; i < 10; i++ {
+		id := w.Watch(key, prefixMatch, 0)
+		if _, ok := idm[id]; ok {
+			t.Errorf("#%d: id %d exists", i, id)
+		}
+		idm[id] = struct{}{}
+
+		s.Put(keyL, val, lease.NoLease)
+
+		resp := <-w.Chan()
+		if resp.WatchID != id {
+			t.Errorf("#%d: watch id in event = %d, want %d", i, resp.WatchID, id)
+		}
+
+		if err := w.Cancel(id); err != nil {
+			t.Error(err)
+		}
+
+		if len(resp.Events) > 1 {
+			if !bytes.Equal(resp.Events[0].Kv.Key, keyL) {
+				t.Errorf("#%d: resp.Events got = %s, want = %s", i, resp.Events[0].Kv.Key, keyL)
+			}
+		}
+	}
+
+	key1, key1L := []byte("foo1"), []byte("foo1bar")
+	s.Put(key1L, val, lease.NoLease)
+
+	// unsynced watchers
+	for i := 10; i < 20; i++ {
```

Comment:
probably reduce this to 5 times? I think this will take a few second to finish? We sync watcher every 0.5 second.